src/models/document/base-document-content.ts

```diff
@@ -1077,5 +1064,148 @@ export const BaseDocumentContentModel = types
+              const width = entry[1] as number;
+              if (width !== undefined && originalId && attributeIdMap[originalId]) {
+                  columnWidths[attributeIdMap[originalId]] = width;
+              }
+            }
+          });
+          tileContent.content = { ...oldContent.content, columnWidths };
+        } else if (tile.tileType === "DataCard") {
+          const oldAttributeId = oldContent.content.selectedSortAttributeId;
+          sharedDataSetEntries.forEach(sharedDataSetEntry => {
+            const originalSharedDataSetId = sharedDataSetEntry.sharedModel.id;
+            const attributeIdMap = updatedSharedModelMap[originalSharedDataSetId].attributeIdMap;
+            if (attributeIdMap[oldAttributeId]) {
+              const content = { ...oldContent.content, selectedSortAttributeId: attributeIdMap[oldAttributeId] };
+              tileContent.content = content;
+            }
+          });
+        }
+
+        // Save the updated tile so we can add it to the document
+        updatedTiles.push({
+          ...tile,
+          newTileId: tileIdMap[tile.tileId],
+          tileContent: JSON.stringify(tileContent)
+        });
+      });
+
+      // Add copied tiles to document
+      const results = self.copyTilesIntoNewRows(updatedTiles, rowIndex);
+
+      // Increment default titles when necessary
+      results.forEach((result, i) => {
+        const newTile = result?.tileId && self.getTile(result.tileId);
+        if (result && newTile) {
+          const tileContentInfo = getTileContentInfo(newTile.content.type);
+          if (tileContentInfo) {
+            const match = titleMatchesDefault(newTile.title, tileContentInfo.titleBase);
+            if (match) {
+              const newTitle = self.getNewTileTitle(newTile.content.type);
+
+              // If the tile title needed to be update, we assume we should also update the data set's name
+              newSharedModelEntries.forEach(sharedModelEntry => {
+                const oldName = sharedModelEntry.sharedModel.dataSet.name;
+                if (oldName === newTile.title) {
+                  sharedModelEntry.sharedModel.dataSet.name = newTitle;
+                }
+              });
+
+              newTile.setTitle(newTitle);
+            }
+          }
+        }
+      });
+
+      // Update tile ids for shared models and add copies to document
+      newSharedModelEntries.forEach(sharedModelEntry => {
+        const updatedTileIds: string[] = sharedModelEntry.tiles.map((oldTile: any) => tileIdMap[oldTile.id]);
+        const updatedSharedModel = { ...sharedModelEntry.sharedModel };
+        if (sharedModelEntry.sharedModel.type === "SharedDataSet") {
+          updatedSharedModel.providerId = tileIdMap[sharedModelEntry.sharedModel.providerId];
+        }
+        const newSharedModelEntry = self.addSharedModel(updatedSharedModel);
+        updatedTileIds.forEach(tileId => newSharedModelEntry.tiles.push(tileId));
```

Comment:
I'm surprised this works. The `sharedModelEntry.tiles` is an array of references. I would have assumed this would only accept Tile instances, but I guess it also accepts ids?

Reply:
At one point I was accessing the `sharedModelManager` and adding the tiles that way, but it wasn't necessary to solve the problem I was facing at the time, and was quite a bit more verbose, so I went back to this way.